### DISCUSSION TITLE
同一バージョンのPGroongaを導入している２つの環境で検索の挙動が異なる

### DISCUSSION THREAD

**Discussion by takasegawa** [pgroonga/pgroonga#273] [Q&A] [answered]
突然の書き込み失礼いたします。初めて投稿させていただきます。
現在下記の2環境にてPGroongaの動作確認をしております。
OS以外の差異が無いよう構築を行ったのですが、それぞれの環境で動作が異なる部分があり、
原因究明のための手がかりとなる情報を頂きたい次第でございます。

以下、環境情報とインデックス、当該事象の概要となります。

■環境
・結合環境
　【OS】Rocky Linux release 8.4 (Green Obsidian)
　【PostgreSQL】14.5（文字コード: UTF8）
　【PGroonga】2.3.4-1.el8（Groonga: 11.1.0）

・システム検証環境（本番相当）
　【OS】Red Hat Enterprise Linux release 8.6 (Ootpa)
　【PostgreSQL】14.5（文字コード: UTF8）
　【PGroonga】2.3.4-1.el8（Groonga: 11.1.0）

■PGroongaインデックス（両環境共通）

> CREATE INDEX idx_pgroonga_text ON test.test_data
> 　USING pgroonga (text)
> 　WITH (plugins='token_filters/stem', tokenizer='TokenBigramSplitSymbol', token_filters='TokenFilterStem')

　※test_data.text のカラムデータ型は「text型」です。

■事象
下記のSQLを実行したところ、結合環境とシステム検証環境で結果の差異がありました。
・実行SQL:
　　select text from test.test_data where text &@~ '|agr';
・検索対象データ：

> |agreement|
> |AGR type reactor|
> |agri-environmental policy|

・想定する結果：
　　検索対象データ3件が取得できること。
・実際の結果：
　　結合環境では3件すべて検索できたが、システム検証環境では「|AGR type reactor|」しか取得できなかった。

■試したこと
システム検証環境のトークナイザーの設定を「TokenBigramSplitSymbolAlpha」に変更したところ、
想定する結果通り3件のデータを取得することができました。
結合環境でも同じ変更を行ったところ、変わらず3件のデータを取得することができました。

■PGroongaログ
以下、実行SQLを実行した際のPGroongaログになります。

・結合テスト環境

> 2022-12-08 15:23:16.758809|n|1323240: grn_init: <11.1.0>
> 2022-12-08 15:23:16.759151|n|1323240: vm.overcommit_memory kernel parameter should be 1: <0>: See INFO level log to resolve this
> 2022-12-08 15:23:16.760013|n|1323240: pgroonga: initialize: <2.3.4>
> 2022-12-08 15:23:16.763839|n|1323239: grn_init: <11.1.0>
> 2022-12-08 15:23:16.764034|n|1323239: vm.overcommit_memory kernel parameter should be 1: <0>: See INFO level log to resolve this
> 2022-12-08 15:23:16.764647|n|1323239: pgroonga: initialize: <2.3.4>
> 2022-12-08 15:23:16.890427|n|1323239: DDL:2147483656:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.890793|n|1323239: DDL:2147483657:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.902273|n|1323239: DDL:2147483661:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.903149|n|1323239: DDL:3221225483:set_source  
> 2022-12-08 15:23:16.906713|n|1323240: DDL:2147483656:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.907379|n|1323240: DDL:2147483657:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.911519|n|1323240: DDL:2147483661:set_normalizers NormalizerAuto
> 2022-12-08 15:23:16.912456|n|1323240: DDL:3221225483:set_source  
> 2022-12-08 15:23:18.499245|n|1323240: grn_fin (0)
> 2022-12-08 15:23:18.503459|n|1323239: grn_fin (0)

・システム検証環境

> 2022-12-08 14:59:18.725734|n|1968604: grn_init: <11.1.0>
> 2022-12-08 14:59:18.725834|n|1968604: vm.overcommit_memory kernel parameter should be 1: <0>: See INFO level log to resolve this
> 2022-12-08 14:59:18.726093|n|1968604: pgroonga: initialize: <2.3.4>
> 2022-12-08 14:59:18.744134|n|1968604: DDL:2147483656:set_normalizers NormalizerAuto
> 2022-12-08 14:59:18.744281|n|1968604: DDL:2147483657:set_normalizers NormalizerAuto

■実行計画
以下、実行SQLにEXPLAIN ANALYZE VERBOSEを付与して実行した結果になります。

・結合テスト環境

> "Gather  (cost=1000.00..24878.48 rows=659 width=18) (actual time=8.985..1717.168 rows=311 loops=1)"
> "  Output: text"
> "  Workers Planned: 2"
> "  Workers Launched: 2"
> "  ->  Parallel Seq Scan on test.test_data  (cost=0.00..23812.58 rows=275 width=18) (actual time=25.376..1529.911 rows=104 loops=3)"
> "        Output: text"
> "        Filter: (test_data.text &@~ '\|agr'::text)"
> "        Rows Removed by Filter: 219518"
> "        Worker 0:  actual time=9.192..1484.263 rows=79 loops=1"
> "        Worker 1:  actual time=58.444..1410.479 rows=112 loops=1"
> "Planning Time: 0.214 ms"
> "Execution Time: 1717.267 ms"

・システム検証環境

> Index Scan using idx_pgroonga_text on test.test_data  (cost=0.00..12.03 rows=659 width=18) (actual time=0.625..0.630 rows=2 loops=1)
>   Output: text
>     Index Cond: (test_data.text &@~ '\|agr'::text)
>     Planning Time: 0.576 ms
>     Execution Time: 0.761 ms

■質問事項
①実行計画を確認した限り、結合環境ではSQL実行時にインデックスが適用されていないことが環境間で検索結果が異なる直接的な原因ではないかと思料しております（なので、本来の挙動としてはシステム検証環境が正しい認識）。
同じSQLを実行しているにも関わらず。結合環境とシステム検証環境でPGroongaインデックスの使用有無に差異が出てしまう原因に何かお心当たりのある方はいらっしゃいますでしょうか。

②トークナイザーを「TokenBigramSplitSymbolAlpha」に変更して想定の検索結果が取得できたのは、クエリ文字列「|agr」が検索対象データから生成されたトークン「|a, ag, gr」に合致するようになったため、という理解で相違ないでしょうか。

③トークナイザーを変更した場合、お作法的に検証すべき観点等ありますでしょうか。
リファレンスを確認した限り、「TokenBigramSplitSymbolAlpha」を採用することで「検索の適合率が下がり、再現率が上がった」と認識しているため、記号や空白等で区切られた文字列を何パターンか用意して検索を行い、想定の結果が得られるかのテストを行おうと考えております。

以上になります。長文失礼いたしました。
いずれか一つでも構いませんので、何らかのご回答を頂けますと大変助かります。

**Comment by komainu8** [comment]
①ですが、ご理解の通り、結合テスト環境ではPGroongaのインデックスが使用されておらずシーケンシャルサーチになっています。一方、システム検証環境ではPGroongaのインデックスが使用されているため検索結果が異なっています。

PGroongaのインデックスが使用されている方が「|AGR type reactor|」しか取得できないのは、トークナイザーが`TokenBigramSplitSymbol`だからです。

PGroongaで`TokenBigramSplitSymbol`を使用した場合、このトークナイザーは記号を2文字区切りでトークナイズしますが、ASCII文字は空白区切りでトークナイズします。

したがって、

`|agreement|` -> `|` / `agreement` / `|`
`|AGR type reactor|` -> `|` / `agr` / `type` / `reactor` / `|`
|agri-environmental policy| -> `|` / `agri` / `-` / `environmental` / `policy` / `|`

とトークナイズされます。
検索キーワードの`|agr`は「`|` / `agr`」とトークナイズされるので、これと一致するトークナイズ結果のレコードは`|AGR type reactor|`しかないため、`|AGR type reactor|`のみがヒットします。

インデックスを使用するかどうかはPostgreSQLのプランナーというやつが入力されたSQLを実行して結果を返すのに最も速いであろう実行方法を予測して実行します。予測なので実際には速くないこともあります。この予測に使うのはPostgreSQLが内部に保持している統計情報というものを使って推測しています。
テーブルのデータ、スキーマが同一であっても、この統計情報が異なっているとPostgerSQLのプランナーは異なる実行方法を採用する可能性があります。

両方の環境で、 `VACUUM ANALYZE` を実行して両方の環境の統計情報を最新にしてから再度、 `EXPLAIN ANALYZE VERBOSE` を実行して結果が変わるかを確認してみてください。

**Comment by komainu8** [comment]
②ですが、ご理解の通りです。`TokenBigramSplitSymbolAlpha`を使うことでアルファベットも2文字区切りでトークナイズされるようになったためです。

**Comment by komainu8** [comment]
③ですが、こちらもご理解の通り、一般に「TokenBigramSplitSymbolAlpha」を採用することで「検索の適合率が下がり、再現率が上がった」状態になります。そのため、「TokenBigramSplitSymbol」よりも検索結果にノイズが多くなっていると思います。（データによってはあまり変化がないかもしれませんが、一般的にはノイズが増えることが多いです。）

ユースケースを把握できていませんが、検索結果が変化するので、「TokenBigramSplitSymbol」を使った検索クエリーについては、再度検索を実行し「TokenBigramSplitSymbolAlpha」を使ったときと、「TokenBigramSplitSymbol」を使ったときの検索結果を比較するのが良いと思います。

比較の結果、使い物にならないくらいノイズが増える（欲しい情報がノイズに埋もれてしまって探し出せない状態）なら、別のアプローチを考える必要がありますし、許容できるノイズで必要な情報がちゃんと取れていることが確認できたのであれば、「TokenBigramSplitSymbolAlpha」を使っても問題ないと判断できると思います。

**Comment by kou** [comment, marked as answer]
`column &@~ (query, weights, index_name)::pgroonga_full_text_search_condition` https://pgroonga.github.io/ja/reference/operators/query-v2.html
という構文を使うとインデックスを使っても使わなくても同じ結果になります。

**Comment by takasegawa** [comment]
@komainu8 様、 @kou 様

早急かつ詳細な回答と解説、誠にありがとうございます。
①の直接的な原因と、②③の内容については疑問が解消いたしました。

①の根本原因調査についてですが、まずはご指摘の通り両環境で下記のコマンドを実行し、結果を確認してみました。
```
VACUUM ANALYZE test_data.text;
EXPLAIN ANALYZE VERBOSE select text from test.test_data where text &@~ '\|agr';
```

結果としては、両環境とも実行計画の内容に変化は見られませんでした。
（結合環境ではインデックスを参照し、システム検証環境ではシーケンシャルサーチを実行）

そこで、kou様からご回答いただいた下記の関数を両環境で実行してみました。
`EXPLAIN ANALYZE VERBOSE select text from test.test_data where text &@~ ('\|agr', null, 'idx_pgroonga_text')::pgroonga_full_text_search_condition;`

その結果、下記の通り両環境でインデックスを参照させることができました。
・結合環境
>Index Scan using idx_pgroonga_text on test.test_data (cost=0.00..16.04 rows=659 width=18) (actual time=1.577..1.594 rows=3 loops=1)
  Output: text 
  Index Cond: (test_data.text &@~ '("\\|agr",,idx_pgroonga_text)'::pgroonga_full_text_search_condition)
Planning Time: 8.907 ms
Execution Time: 1.992 ms

・システム検証環境
>Index Scan using idx_pgroonga_text on test.test_data (cost=0.00..12.03 rows=659 width=18) (actual time=0.644..0.649 rows=2 loops=1)
  Output: text
  Index Cond: (test_data.text &@~ '("\\|agr",,idx_pgroonga_text)'::pgroonga_full_text_search_condition)
Planning Time: 0.601 ms
Execution Time: 0.770 ms

上記の実装により、両環境で意図した挙動を実現することはできそうです。
結合環境にてインデックスが使用されなかった原因について、他に確認できる観点等ありますでしょうか。
無ければ一旦上記対応で実装上問題無いか検証してみようと思います。

**Reply by komainu8** [reply]
片方だけパラレルクエリーが使われているので動けるワーカー数に違いがあるとか（違いはなさそうな気がしますが。。。）、または、再現できるデータを頂ければなにかわかるかもしれませんが、解決には時間がかかりそうな気がしています。

もとの書き方で実行しなければならないという強い理由がなければ、もとの書き方の問題を解決するのに時間を使うよりは、 `column &@~ (query, weights, index_name)::pgroonga_full_text_search_condition` の書き方のほうがシーケンシャルサーチ(PGroongaのインデックスが使われない)の時でもPGroongaのインデックスに指定した検索オプション(トークナイザーやトークンフィルターなど)を使えるので `column &@~ (query, weights, index_name)::pgroonga_full_text_search_condition` の書き方で問題ないかを確認して、問題ないならこっちを使う。という方針のほうが良いかと思います。

**Reply by takasegawa** [reply]
上記承知いたしました。
対応期間もあまり残されておらず、関数を使うにあたりコーディング上の規約制限もありませんので、今回は`column &@~ (query, weights, index_name)::pgroonga_full_text_search_condition`を使う方向で一旦検討しようと思います。
まずは上記をソースに組み込み、問題なく動くかを検証する予定です。

早急な回答、アドバイスを賜り、誠にありがとうございました。

**Reply by komainu8** [reply]
また何かあればお気軽にご質問いただければと思います。

**Reply by kou** [reply]
`SHOW ALL`で実際に設定されている値の一覧を見れるんですが、結合環境での結果とシステム検証環境での結果を比較するとなにか違いがあるのかもしれません。
